app/utils/smartapi/data_processor.py

```diff
@@ -35,3 +45,172 @@ def process_smart_api_stock_data(
+    which is either 9:15 or time in given start_datetime to end time which is either 15:29 or time in given end_datetime.
+
+    Parameters:
+    -----------
+    current_date: ``datetime``
+        The date for which to find the range of timestamps.
+    start_datetime: ``str``
+         The initial datetime from which historical stock data should be retrieved.
+    end_datetime: ``str``
+        The final datetime up to which historical stock data should be retrieved.
+    interval: ``CandlestickInterval``
+        The interval of the candlestick
+
+    Return:
+    -------
+    List[str]
+        The possible time range for given current date.
+    """
+    if interval.name == "ONE_DAY":
+        return [f"{current_date.strftime('%Y-%m-%d')}T00:00:00+05:30"]
+    start_time = time(9, 15)
+    end_time = time(15, 29)
+    if start_datetime.date() == current_date.date():
+        start_time = max(start_time, start_datetime.time())
+    if end_datetime.date() == current_date.date():
+        end_time = min(end_time, end_datetime.time())
+    # Create a datetime range for every minute from 9:15 to 15:29 on the particular day
+    time_range = pd.date_range(
+        start=f"{current_date.date()} {start_time.strftime('%H:%M')}:00+05:30",
+        end=f"{current_date.date()} {end_time.strftime('%H:%M')}:00+05:30",
+        freq=f"{interval.value[0]}min",
+    )
+
+    # Convert the datetime range to a list of timestamps of string type
+    time_range_list = [
+        timestamp.strftime("%Y-%m-%dT%H:%M:%S+05:30") for timestamp in time_range
+    ]
+
+    return time_range_list
+
+
+def get_missing_timestamps(
+    historical_stock_data: List[List[str]],
+    stock_symbol: str,
+    interval: CandlestickInterval,
+    start_datetime: datetime,
+    end_datetime: datetime,
+) -> List[str]:
+    """Finds the missing timestamps of given interval in the historical stock data
+    of a stock between start_datetime and end_datetime.
+    Parameters:
+    -----------
+    historical_stock_data: ``List[List[str]]``
+        Available historical stock data points.
+    stock_symbol: ``str``
+        The symbol of the stock.
+    interval: ``CandlestickInterval``
+        Candlestick interval.
+    start_datetime: ``datetime``
+         The initial datetime from which to find timestamps of missing data points.
+    end_datetime: ``datetime``
+        The final datetime up to which to find timestamps of missing data points.
+
+    Return:
+    -------
+    List[str]
+        List of missing timestamps in given historical stock data.
+    """
+    df = pd.DataFrame(historical_stock_data)
+    full_datetime_range = []
+    missing_timestamps = []
+    try:
+        start_datetime = check_data_availability(
+            start_datetime, end_datetime, stock_symbol.split("-")[0], interval
+        )
+        open_dates = find_open_market_days(start_datetime, end_datetime)
+        for open_date in open_dates:
```

Comment:
Use list comprehension here as well as you are not using `full_datetime_range` anywhere except `np.setdiff1d`.

Reply:
changed for loop to list comprehension and itr tools.